[Test] [history server] [collector] Add collector e2e tests

JiangJiaWei1103 opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

In this PR, we add a test suite to validate the behavior of the collector (including both logs and events), making sure it works seamlessly together with the storage component (e.g., MinIO) in real-world use cases.

### Test Case Description

The following summarize two main test cases we add in this PR:

#### 1. Happy path: Logs and events should be uploaded to S3 on deletion

- **Goal**: Verify `logs` and `node_events` are successfully uploaded to S3 upon the Ray cluster is deleted
- Process overview
  - Prepare test env
    - Deploy a Ray cluster, check collector sidecar, and check the automatically created S3 bucket
  - Submit a Ray job to the existing Ray cluster
  - Retrieve `sessionID` and `nodeID` for verifying log and event persistence
  - Delete the Ray cluster to trigger log uploading and event flushing
  - Check `logs` and `node_events` object existence and **content** (by file size)
  - Delete S3 bucket to ensure isolation

#### 2. Single session single node logs should be uploaded to S3 during runtime

- **Goal**: Verify log snapshots under `/tmp/ray/prev-logs` are uploaded to S3 and events are flushed during runtime
- Process overview
  - Prepare test env
    - Deploy a Ray cluster, check collector sidecar, and check the automatically created S3 bucket
  - Submit a Ray job to the existing Ray cluster
  - Retrieve `sessionID` and `nodeID` for verifying log and event persistence
  - Kill the main process of the `ray-head` container to trigger restart
    - Logs are moved from `/tmp/ray/session_latest` to `/tmp/ray/prev-logs` by the startup command
    - Events are flushed since the `/tmp/ray/raylet_node_id` changes
  - Check `logs` and `node_events` object existence and **content** (by file size)
  - Delete the Ray cluster to clear resources
  - Delete S3 bucket to ensure isolation

## Related issue number

Closes https://github.com/ray-project/kuberay/issues/4275.

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [x] Manual tests
  - [ ] This PR is not tested :(

JiangJiaWei1103 commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,174 @@
+	minioNamespace    = "minio-dev"
+	minioManifestPath = "../../config/minio.yaml"
+	minioUsername     = "minioadmin"
+	minioSecret       = "minioadmin"
```
This is named as `Secret` since `Password` failed the gitleaks detection.

machichima commented on historyserver/pkg/collector/logcollector/runtime/logcollector/collector.go:
```diff
@@ -664,10 +664,10 @@ func (r *RayLogHandler) processPrevLogsDir(sessionNodeDir string) {
 		if _, err := os.Stat(logsDir); os.IsNotExist(err) {
 			logrus.Warnf("Logs directory does not exist: %s", logsDir)
-			time.Sleep(time.Millisecond * 10)
+			time.Sleep(time.Second * 1)
```
Probably we can use `fsnotify` like what we did in other place in this file. Can be a follow-up

machichima commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,351 @@
+// testPrevLogsRuntimeUpload verifies logs under /tmp/ray/prev-logs are uploaded during runtime.
+// This makes sure WatchPrevLogsLoops processes logs as they appear.
+//
+// NOTE: For now, logs under /tmp/ray/session_latest are moved to /tmp/ray/prev-logs explicitly.
```
Could you add more details saying that this test move logs to `/tmp/ray/prev-logs`? And explain why we move it?

machichima commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,351 @@
+	}, TestTimeoutMedium).Should(Succeed(), "Failed to move logs to prev-logs directory")
+
+	LogWithTimestamp(test.T(), "Waiting for collector to detect and process prev-logs")
+	time.Sleep(3 * time.Second)
```
It's better to prevent using `time.Sleep` in the test to prevent flakiness. Could we use `g.Eventually` here to check if the collector process the prev-logs?
Or can we remove this and directly use `g.Eventually` below?

JiangJiaWei1103 replied:
Fixed at https://github.com/ray-project/kuberay/pull/4308/commits/6b0ffdc160b30c54d98273aaf74edb91abb5b4c9. Thanks!

machichima commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,351 @@
+		})
+		// TODO(jwj): Add err handling for ListObjectsV2.
+		return aws.Int64Value(objects.KeyCount)
+	}, TestTimeoutMedium).Should(BeNumerically(">", 0))
```
Can simply do this for handling error

```go
g.Eventually(func(gg Gomega) {
      objects, err := s3Client.ListObjectsV2(&s3.ListObjectsV2Input{
          Bucket: aws.String(s3BucketName),
      })
      gg.Expect(err).NotTo(HaveOccurred())
      gg.Expect(aws.Int64Value(objects.KeyCount)).To(BeNumerically(">", 0))
  }, TestTimeoutMedium).Should(Succeed())
```

JiangJiaWei1103 replied:
Fixed at https://github.com/ray-project/kuberay/pull/4308/commits/b83b05c5ffe289f88f082565d4be7202b5c6bf72 to check both logs and node_events with Gomega multiple assertions. Thanks!

machichima commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,351 @@
+	_, err = s3Client.HeadBucket(&s3.HeadBucketInput{
+		Bucket: aws.String(s3BucketName),
+	})
+	g.Expect(err).NotTo(HaveOccurred())
```
Maybe we can extract this into a setup function, and used in here and in `testPrevLogsRuntimeUpload`?

JiangJiaWei1103 replied:
Works for me! Extracted into `prepareTestEnv` at https://github.com/ray-project/kuberay/pull/4308/commits/cff5d231239b93893062b75e9aeecc70e078f73c, which can be reused in newly added test cases in the future. Thanks!

JiangJiaWei1103 commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,146 @@
+	//  ...
+}
+
+func testLogUploadOnDeletion(test Test, g *WithT, namespace *corev1.Namespace) {
```
We add more test cases summarized in the PR description.

JiangJiaWei1103 commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,458 @@
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
```
Thank you for pointing this out!

Since the AWS SDK upgrade is already tracked in this [issue](https://github.com/ray-project/kuberay/issues/4280)￼, I suggest we first make these e2e tests work with the current setup and then handle the SDK migration in a follow-up PR. WDYT?

cc @Future-Outlier

JiangJiaWei1103 commented:
Hi @Future-Outlier,

Regarding the order of test functions and helpers, I refer to how e2e tests of `ray-operator` structure their codes and observe that common helpers are put under `support/` so they can be shared among different test files. Considering we only have one e2e test file for the collector, would it make sense to keep the current order:

1. Constants
2. Test function
3. Test helpers (internal test cases)
4. Setup/Teardown helpers
5. Utilies (the most general helpers like `getContainerStatusByName`)

Thanks a lot !

JiangJiaWei1103 commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,449 @@
+				hasLogsDir = true
+			case "node_events":
+				hasNodeEventsDir = true
+			}
```
We've modified this check to make sure both `logs/` and `node_events/` must be pushed to S3.

win5923 commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,434 @@
+	g.Eventually(func() error {
+		_, err := GetRayCluster(test, rayCluster.Namespace, rayCluster.Name)
+		return err
+	}, TestTimeoutMedium).Should(WithTransform(k8serrors.IsNotFound, BeTrue()))
```
Is this trigger `processSessionLatestLogs` to upload logs? If so, can you add more comments on the testLogAndEventUploadOnDeletion to describe? Thanks!
https://github.com/ray-project/kuberay/blob/7791a8786861818f0cebcce381ef221436a0fa4d/historyserver/pkg/collector/logcollector/runtime/logcollector/collector.go#L70-L83

win5923 commented on historyserver/test/e2e/collector_test.go:
```diff
@@ -1 +1,434 @@
+
+	// Add other test cases below.
+	// ...
+}
```
Not sure why I got this error in my local e2e test:
```
$ KUBERAY_TEST_TIMEOUT_SHORT=1m KUBERAY_TEST_TIMEOUT_MEDIUM=5m KUBERAY_TEST_TIMEOUT_LONG=10m go test -timeout 30m -v ./test/e2e
...

    collector_test.go:240: [2025-12-31T08:01:18Z] Deleting S3 bucket ray-historyserver
    collector_test.go:280: [2025-12-31T08:01:18Z] Deleted S3 bucket ray-historyserver successfully
    test.go:114: [2025-12-31T08:01:18Z] Retrieving Pod Container test-ns-6bvs2/ray-job-4hs4q/ray-job-submitter logs
    test.go:102: [2025-12-31T08:01:18Z] Creating ephemeral output directory as KUBERAY_TEST_OUTPUT_DIR env variable is unset
    test.go:105: [2025-12-31T08:01:18Z] Output directory has been created at: /tmp/TestCollector3649020351/001
    test.go:114: [2025-12-31T08:01:18Z] Retrieving Pod Container test-ns-6bvs2/raycluster-historyserver-head-xz24v/ray-head logs
    test.go:114: [2025-12-31T08:01:18Z] Retrieving Pod Container test-ns-6bvs2/raycluster-historyserver-head-xz24v/collector logs
    test.go:114: [2025-12-31T08:01:18Z] Retrieving Pod Container test-ns-6r9rh/ray-job-gpjjg/ray-job-submitter logs
    test.go:114: 
                Error Trace:    /home/ubuntu/go/pkg/mod/github.com/ray-project/kuberay/ray-operator@v1.5.1/test/support/utils.go:24
                                                        /home/ubuntu/go/pkg/mod/github.com/ray-project/kuberay/ray-operator@v1.5.1/test/support/core.go:68
                                                        /home/ubuntu/go/pkg/mod/github.com/ray-project/kuberay/ray-operator@v1.5.1/test/support/core.go:44
                                                        /home/ubuntu/go/pkg/mod/github.com/ray-project/kuberay/ray-operator@v1.5.1/test/support/test.go:114
                                                        /usr/local/go/src/testing/testing.go:1308
                                                        /usr/local/go/src/testing/testing.go:1572
                                                        /usr/local/go/src/testing/testing.go:1928
                Error:          Received unexpected error:
                                open /tmp/TestCollector3649020351/001/pod-ray-job-gpjjg-ray-job-submitter.log: no such file or directory
                Test:           TestCollector
--- FAIL: TestCollector (84.50s)
    --- PASS: TestCollector/Happy_path:_Logs_and_events_should_be_uploaded_to_S3_on_deletion (36.55s)
    --- PASS: TestCollector/Single_session_single_node_logs_and_events_should_be_uploaded_to_S3_during_runtime (46.50s)
FAIL
FAIL    github.com/ray-project/kuberay/historyserver/test/e2e   84.545s
FAIL
```

Future-Outlier replied:
I have the same problem too.

win5923 replied:
Maybe it's because two calls to `NewTestNamespace()` both register cleanup on the same test object.

https://github.com/ray-project/kuberay/blob/7791a8786861818f0cebcce381ef221436a0fa4d/ray-operator/test/support/test.go#L110-L119

To avoid this, each subtest should create its own test context and namespace
```
func TestCollector(t *testing.T) {
      s3Client := ensureS3Client(t)

      t.Run("Happy path: Logs and events should be uploaded to S3 on deletion", func(t *testing.T) {
          test := With(t)
          g := NewWithT(t)
          namespace := test.NewTestNamespace()

          testLogAndEventUploadOnDeletion(test, g, namespace, s3Client)
      })

      t.Run("Single session single node logs and events should be uploaded to S3 during runtime", func(t *testing.T) {
          test := With(t)
          g := NewWithT(t)
          namespace := test.NewTestNamespace()

          testLogAndEventUploadDuringRuntime(test, g, namespace, s3Client)
      })
  }
```

JiangJiaWei1103 replied:
Thanks for pointing this out. Fixed at https://github.com/ray-project/kuberay/pull/4308/commits/831866d16359d0cf1331fe68092b4cb58cf761b9.

I was able to track down why my local e2e tests never hit this error. It turns out I already have `KUBERAY_TEST_OUTPUT_DIR` set in my local environment, which is also why CI doesn’t fail. This var is explicitly set there as well:

https://github.com/ray-project/kuberay/blob/831866d16359d0cf1331fe68092b4cb58cf761b9/.buildkite/test-historyserver-e2e.yml#L20

When this env var isn't set, the test object creates a temporary directory the first time `storeAllPodLogs` is called during `Cleanup`:

https://github.com/ray-project/kuberay/blob/7791a8786861818f0cebcce381ef221436a0fa4d/ray-operator/test/support/test.go#L103
https://github.com/ray-project/kuberay/blob/7791a8786861818f0cebcce381ef221436a0fa4d/ray-operator/test/support/test.go#L114

I believe this explains the `no such file or directory` error: two subtests end up sharing the same temporary output directory, which leads to conflicts when it is created or cleaned up.

Explicitly creating the output temp dir before running subtests also works:

```go
func TestCollector(t *testing.T) {
	test := With(t)
	g := NewWithT(t)
	
	# Create the temporary dir.
    _ = test.OutputDir()

	// Create an isolated Kubernetes namespace.
	namespace := test.NewTestNamespace()

	// Share a single S3 client among subtests.
	s3Client := ensureS3Client(test, g)

	t.Run("Happy path: Logs and events should be uploaded to S3 on deletion", func(t *testing.T) {
		testLogAndEventUploadOnDeletion(test, g, namespace, s3Client)
	})

	namespace = test.NewTestNamespace() // Separate namespace to prevent resource contention.
	t.Run("Single session single node logs and events should be uploaded to S3 during runtime", func(t *testing.T) {
		testLogAndEventUploadDuringRuntime(test, g, namespace, s3Client)
	})

	// Add other test cases below.
	// ...
}
```

Future-Outlier reviewed (commented):
1.	Add TODO comments; use a for loop.
2.	Add two comments for both tests.
•	Ref: ray-operator/controllers/ray/rayservice_controller.go
•	Topic: RayService incremental upgrade
3.	 add comments when using Kill PID 1 — note that only on Kubernetes 1.28+ does an OOM reliably trigger a container restart.
4.	Check whether the second test also needs this:

```go
err := test.Client().Ray().RayV1().
    RayClusters(rayCluster.Namespace).
    Delete(test.Ctx(), rayCluster.Name, metav1.DeleteOptions{})
g.Expect(err).NotTo(HaveOccurred())
g.Eventually(func() error {
    _, err = GetRayCluster(test, rayCluster.Namespace, rayCluster.Name)
    return err
}, TestTimeoutMedium).Should(WithTransform(k8serrors.IsNotFound, BeTrue()))
```

5.  Check these directories exist in the second test: 
	


```zsh

(base) ray@raycluster-historyserver-head:/tmp/ray/prev-logs$ ls
session_2025-12-31_19-22-14_425845_1  session_2025-12-31_19-26-36_646888_1
```

JiangJiaWei1103 commented:
> 1. Add TODO comments; use a for loop.
> 2. Add two comments for both tests.
>    •	Ref: ray-operator/controllers/ray/rayservice_controller.go
>    •	Topic: RayService incremental upgrade
> 3. add comments when using Kill PID 1 — note that only on Kubernetes 1.28+ does an OOM reliably trigger a container restart.
> 4. Check whether the second test also needs this:
> 
> ```go
> err := test.Client().Ray().RayV1().
>     RayClusters(rayCluster.Namespace).
>     Delete(test.Ctx(), rayCluster.Name, metav1.DeleteOptions{})
> g.Expect(err).NotTo(HaveOccurred())
> g.Eventually(func() error {
>     _, err = GetRayCluster(test, rayCluster.Namespace, rayCluster.Name)
>     return err
> }, TestTimeoutMedium).Should(WithTransform(k8serrors.IsNotFound, BeTrue()))
> ```
> 
> 5. Check these directories exist in the second test:
> 
> ```shell
> (base) ray@raycluster-historyserver-head:/tmp/ray/prev-logs$ ls
> session_2025-12-31_19-22-14_425845_1  session_2025-12-31_19-26-36_646888_1
> ```

Hi @Future-Outlier, we’ve addressed and applied all of your suggestions. Thank you very much for the thoughtful review. I learned a lot from your feedback!

Here’s a summary of the changes:
1. Iterated over subtests, making it easier to add new cases in the future
2. Added clearer, step-by-step descriptions for each test case so the intent is easier to understand
    - We now use observable behavior (the how) to explain the underlying intent (the why).
3. Clarified that `kill 1` is used to simulate an `OOMKilled` event (Kubernetes 1.28+ with cgroups v2 enabled)
4. Removed redundant cluster cleanup, since a cleanup hook (test namespace deletion) is already registered [here](https://github.com/ray-project/kuberay/blob/7791a8786861818f0cebcce381ef221436a0fa4d/ray-operator/test/support/test.go#L116)￼
5. Verified that the old session dir (via `sessionID`) is moved to both `prev-logs` (by the ray-head startup command) and `persist-complete-logs` (after logs are processed and uploaded to S3).
6. Refactored `verifyS3SessionDirs` so `logs` and `node_events` are checked using the same logic

Thanks again for your guidance!

Future-Outlier reviewed (approved):
LGTM, thank you!

<img width="1548" height="171" alt="image" src="https://github.com/user-attachments/assets/33409bc4-b1cb-4be9-b26b-1196fc0ea4d3" />
